File: Trees/872 Leaf-Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> ans1;
    vector<int> ans2;
    void Check(TreeNode* root,vector<int> &ans)
    {
        if(root==NULL)
            return;
        if(root->left==NULL && root->right==NULL)
            ans.push_back(root->val);
        Check(root->left,ans);
        Check(root->right,ans);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) 
    {
        Check(root1,ans1);
        Check(root2,ans2);
        return ans1==ans2;
    }
};

File: Trees/872 Leaf-Similar Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def traverse1(self,root):
        if not root:
            return 
        if(root.left==None and root.right==None):
            self.s1+=str(root.val)+" "
        self.traverse1(root.left)
        self.traverse1(root.right)

    def traverse2(self,root):
        if not root:
            return 
        if(root.left==None and root.right==None):
            self.s2+=str(root.val)+" "
        self.traverse2(root.left)
        self.traverse2(root.right)
        
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        self.s1=""
        self.s2=""
        self.traverse1(root1)
        self.traverse2(root2)
        return self.s1==self.s2
